src/cargo/core/source.rs

```diff
@@ -224,9 +226,39 @@ impl SourceId {
         SourceId::for_registry(&url)
     }
 
+    pub fn alt_registry(config: &Config, key: &str) -> CargoResult<SourceId> {
+        let registries = config.get_table("registries")?;
+        match registries.as_ref().and_then(|registries| registries.val.get(key)) {
+            Some(registry)  => {
+                let index = match *registry {
+                    CV::String(ref s, _)        => s,
+                    CV::Table(ref registry, _)  => {
+                        match registry.get("index") {
+                            Some(index) => index.string(&format!("registries.{}", key))?.0,
+                            None => return Err(format!("No index for registry `{}`", key).into()),
+                        }
+                    }
+                    _ => registry.expected("string or table", &format!("registries.{}", key))?
+                };
+
+                Ok(SourceId::new(Kind::AlternateRegistry(key.to_owned()), index.to_url()?))
+            }
+            None => Err(format!("Required unknown registry source: `{}`", key).into())
+        }
+    }
+
     pub fn url(&self) -> &Url {
         &self.inner.url
     }
+
+    pub fn display_registry(&self) -> String {
+        if let Kind::AlternateRegistry(ref key) = self.inner.kind {
+            format!("alternate registry `{}` ({})", self.url(), key)
+        } else {
+            format!("registry `{}`", self.url())
```

Comment:
:bike::house: : name "alternate registry" does not sound like the best one to present to the user. I can imagine that in corporate environment "alternative" registry is thought of as a primary one. 

Perhaps the messages could look like this?

```
Updating crates.io registry
Updating registry https://my-company.com/cargo-registry
```

Hm, now that I've written it down, perhaps we even should use symbolic names for registries for user-facing messages, and not URLs?

Reply:
Yeah it seems reasonable to me to basically universally refer to registries through their "short identifier", e.g. `crates-io` or whatever's in `.cargo/config`

Reply:
I agree with identifying them using the symbolic name the user has given it, rather than the index URL.

Don't have a strong opinion about how/if to distinguish crates.io from other registries.

Reply:
I'm going to with:

```rust
registry `crates-io`
registry `foobar` (https://github.com/myco/foobar-index)
```

That is, hardcode non alt registry sources to crates-io.

Reply:
However, currently we say:

```
registry `https://github.com/rust-lang/crates.io-index
```

I think giving crates-io instead of the index URL is a nicer UX though.

Reply:
Actually, we show the registry URL because it could be a local path. Gonna keep showing it in all cases, but call the main registry `crates-io`.

Reply:
> I'm going to with:

Looks great! 

> I think giving crates-io instead of the index URL is a nicer UX though.

Yes!

> Actually, we show the registry URL because it could be a local path. Gonna keep showing it in all cases, but call the main registry crates-io.

That may just be an artifact of the tests, but I'm fine either way